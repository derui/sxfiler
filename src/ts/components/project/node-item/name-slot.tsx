import classNames from "classnames";
import * as React from "react";

// tslint:disable-next-line
const styles = require("./node-item.module.scss");

interface Prop {
  isDirectory: boolean;
  isSymlink: boolean;
  name: string;
}

const NodeName: React.FC<Prop> = ({ isDirectory, name, isSymlink }) => {
  const className = classNames(styles.name, {
    [styles.nameDirectory]: isDirectory && !isSymlink,
    [styles.nameSymlink]: isSymlink,
  });

  return <span className={className}>{name}</span>;
};

export default NodeName;
