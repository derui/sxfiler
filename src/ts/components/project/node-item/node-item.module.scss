@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  composes: listItem from "../../ui/list-item/list-item.module.scss";

  padding-bottom: $baseSpace;

  &:first-of-type {
    padding-top: $baseSpace;
  }

  font-family: monospace;
  color: $sol_base1;
  white-space: nowrap;

  font-size: 1rem;
  border-top: 1px solid transparent;
  border-bottom: 1px solid transparent;
}

.rootSelected {
  border-bottom: 1px solid $sol_red;
}

.rootMarked {
  background-color: rgba($sol_blue, 0.25);
}

.timestamp {
  flex: 0 1 auto;
  padding: 0 $smallSpace;
  text-align: right;

  white-space: nowrap;
}

.name {
  flex: 1 0 auto;
  padding: 0;
  font-weight: bold;
}

.name.nameDirectory {
  color: $sol_yellow;
}

.name.nameSymlink {
  color: $sol_orange;
}

.size {
  flex: 0 1 auto;
  padding: 0 $baseSpace;
  text-align: right;

  white-space: pre;
}

.mode {
  flex: 0 1 auto;
  padding: 0 $smallSpace;

  white-space: nowrap;
}
