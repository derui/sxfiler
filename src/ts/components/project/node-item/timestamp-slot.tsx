import * as React from "react";

interface Prop {
  timestamp: Date;
}

/**
 * format Date to display in item
 */
function format(timestamp: Date): string {
  const year = `${timestamp.getFullYear()}`.padStart(4, "0");
  const month = `${timestamp.getMonth() + 1}`.padStart(2, "0");
  const date = `${timestamp.getDate()}`.padStart(2, "0");

  return `${year}/${month}/${date}`;
}

const NodeTimestamp: React.FC<Prop> = prop => {
  const date = format(prop.timestamp);
  return <span className="fp-FileItem_Timestamp">{date}</span>;
};

export default NodeTimestamp;
