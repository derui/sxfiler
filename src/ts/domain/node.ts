// define type and operations for node
import { FileStat } from "./file-stat";

export interface Node {
  id: string;
  name: string;
  stat: FileStat;
  parentDirectory: string;
  linkPath?: string;
}
