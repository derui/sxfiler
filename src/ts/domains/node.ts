// define type and operations for node
import { FileStat } from "./file-stat";

interface FactoryArg {
  id: string;
  name: string;
  stat: FileStat;
  parentDirectory: string;
  linkPath?: string;
}

/**
 * create node
 * @param args
 */
export function create(args: FactoryArg) {
  return new NodeImpl(args.id, args.name, args.stat, args.parentDirectory, args.linkPath);
}

export interface Node {
  readonly id: string;
  readonly name: string;
  readonly stat: FileStat;
  readonly parentDirectory: string;
  readonly linkPath?: string;
}

class NodeImpl implements Node {
  /**
   * create new Node
   * @param id id of node
   * @param name file or directory name of node
   * @param stat stats of node
   * @param parentDirectory parent directory of node
   * @param linkPath link target path if node is link
   */
  constructor(
    public readonly id: string,
    public readonly name: string,
    public readonly stat: FileStat,
    public readonly parentDirectory: string,
    public readonly linkPath?: string
  ) {}
}
