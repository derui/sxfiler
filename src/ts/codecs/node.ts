import { NodeObject, createNode } from "../domains/node";
import { encode as encodeFileStat, TypeOnRPC as FileStatOnRPC } from "./file-stat";

// define codec that is between filer domain and RPC

export type NodeOnRPC = {
  id: string;
  parent: string;
  name: string;
  stat: FileStatOnRPC;
  linkPath?: string;
};

/**
   encode node object from RPC to frontend domain.

   @param obj JSON representation for node
   @return Node object
 */
export const encode = (obj: NodeOnRPC & { marked: boolean }): NodeObject => {
  const stat = encodeFileStat(obj.stat);

  return createNode({ ...obj, stat, parentDirectory: obj.parent });
};
