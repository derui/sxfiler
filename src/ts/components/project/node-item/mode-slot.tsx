import * as React from "react";
import { CapabilityObject } from "../../../domains/capability";
import { ModeObject } from "../../../domains/mode";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./node-item.module.scss");

function capabilityToString(cap: CapabilityObject) {
  const readable = cap.readable ? "r" : "-";
  const writable = cap.writable ? "w" : "-";
  const executable = cap.executable ? "x" : "-";

  return `${readable}${writable}${executable}`;
}

/**
 * convert mode to string
 */
function modeToString(mode: ModeObject, isDirectory: boolean, isSymlink: boolean): string {
  let state = " ";

  if (isDirectory && !isSymlink) {
    state = "d";
  } else if (isSymlink) {
    state = "l";
  }

  const owner = capabilityToString(mode.owner);
  const group = capabilityToString(mode.group);
  const others = capabilityToString(mode.others);

  return `${state}${owner}${group}${others}`;
}

interface Prop {
  mode: ModeObject;
  isDirectory: boolean;
  isSymlink: boolean;
}

const NodeMode: React.FC<Prop> = prop => {
  const data = modeToString(prop.mode, prop.isDirectory, prop.isSymlink);

  return <span className={styles.mode}>{data}</span>;
};

export default NodeMode;
