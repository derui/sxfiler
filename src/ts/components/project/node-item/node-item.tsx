import * as React from "react";
import * as Domain from "../../../domains/node";
import * as ListItem from "../../ui/list-item/list-item";

import Mode from "./mode-slot";
import Name from "./name-slot";
import Size from "./size-slot";
import Timestamp from "./timestamp-slot";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./node-item.module.scss");

export interface Props {
  item: Domain.Node;
  selected: boolean;
}

const Element = ListItem.createComponent();

export class Component extends React.PureComponent<Props> {
  public render() {
    const { item, selected } = this.props;

    return (
      <Element className={styles.nodeItem} aria-selected={selected} data-marked={item.marked}>
        <Mode key="mode" mode={item.stat.mode} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
        <Timestamp key="timestamp" timestamp={item.stat.mtime} />
        <Size key="size" size={item.stat.sizeAsBigInt} />
        <Name key="name" name={item.name} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
      </Element>
    );
  }
}
