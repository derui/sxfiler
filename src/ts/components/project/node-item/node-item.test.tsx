import * as React from "react";
import renderer from "react-test-renderer";

import { create } from "../../../domains/node";

import FileStatFactory from "../../../domains/file-stat-factory";
import { Component as T } from "./node-item";

function makeNode(marked: boolean, isDirectory = false, isSymlink = false) {
  return create({
    id: "node",
    name: "file.ext",
    marked,
    stat: FileStatFactory.create({
      mode: "644",
      uid: 1000,
      gid: 1000,
      atime: "0",
      ctime: "0",
      mtime: "0",
      size: "10",
      isDirectory: isDirectory,
      isFile: !isDirectory && !isSymlink,
      isSymlink: isSymlink,
    }),
    parentDirectory: "/",
  });
}

describe("Project", () => {
  describe("Node Item", () => {
    it("should print correctly", () => {
      const node = makeNode(false);
      const tree = renderer.create(<T item={node} selected={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should add class correctly when marked", () => {
      const node = makeNode(true);
      const tree = renderer.create(<T item={node} selected={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should add class correctly when selected", () => {
      const node = makeNode(false);
      const tree = renderer.create(<T item={node} selected={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should add class correctly when selected and marked", () => {
      const node = makeNode(true);
      const tree = renderer.create(<T item={node} selected={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
