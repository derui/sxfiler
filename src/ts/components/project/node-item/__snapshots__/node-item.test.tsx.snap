// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node Item should add class correctly when marked 1`] = `
<div
  aria-selected={false}
  className="nodeItem"
  data-marked={true}
  role="listitem"
>
  <span
    className="mode"
  >
    =-w----r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <span
    className="size"
  >
    10.0B
  </span>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;

exports[`Project Node Item should add class correctly when selected 1`] = `
<div
  aria-selected={true}
  className="nodeItem"
  data-marked={false}
  role="listitem"
>
  <span
    className="mode"
  >
    =-w----r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <span
    className="size"
  >
    10.0B
  </span>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;

exports[`Project Node Item should add class correctly when selected and marked 1`] = `
<div
  aria-selected={true}
  className="nodeItem"
  data-marked={true}
  role="listitem"
>
  <span
    className="mode"
  >
    =-w----r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <span
    className="size"
  >
    10.0B
  </span>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;

exports[`Project Node Item should print correctly 1`] = `
<div
  aria-selected={false}
  className="nodeItem"
  data-marked={false}
  role="listitem"
>
  <span
    className="mode"
  >
    =-w----r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <span
    className="size"
  >
    10.0B
  </span>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;
